### PR TITLE
Timers use single executor

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

followup to #10118

Changelog:
- update: use main execution thread for scheduling timer events

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: lib/java/ydoc-server/src/main/java/org/enso/ydoc/Ydoc.java [hunk 1]
```diff
@@ -43,15 +43,15 @@ public static class Builder {
...
 
-    private ExecutorService executor;
+    private ScheduledExecutorService executor;
     private ParserPolyfill parser;
     private Context.Builder contextBuilder;
     private String hostname;
     private int port = -1;
 
-    public Builder() {}
+    private Builder() {}
```
Comment (JaroslavTulach) [inline lib/java/ydoc-server/src/main/java/org/enso/ydoc/Ydoc.java:52]: Replacing constructor with a factory method (`Ydoc.builder()`) is good.

File: lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/Timers.java [hunk 2]
```diff
@@ -92,7 +84,17 @@ public Object execute(Value... arguments) {
     };
   }
 
-  private Runnable execute(Value func, Object[] args) {
+  private Runnable runnable(Value func, Object[] args) {
     return () -> executor.execute(() -> func.executeVoid(args));
```
Comment (JaroslavTulach) [inline lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/Timers.java]: I guess this is wrong. It should be just `return () -> func.executeVoid(args)`.

File: lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/Timers.java [hunk 3]
```diff
@@ -92,7 +84,17 @@ public Object execute(Value... arguments) {
...
   }
+
+  /**
+   * Applies Node.js constraints on a delay value.
+   *
+   * @param delay the input delay value.
+   * @return the delay value bounded to the permitted range.
+   */
+  private long normalizedDelay(long delay) {
+    return Math.min(MAXIMUM_DELAY, Math.max(MINIMUM_DELAY, delay));
```
Comment (JaroslavTulach) [inline lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/Timers.java:94]: What are "Node.js constraints"? According to [MDN setInterval](https://developer.mozilla.org/en-US/docs/Web/API/setInterval) defaults to zero if not specified. 

I'd say `MAXIMUM_DELAY` being `1` is _too much_ and it changes semantics. We want to support _delay 0_ to run as soon as possible once current (event queue) task is over.

**Review by JaroslavTulach — APPROVED** [review]
TCR: explain what's happening in the `runnable` method. I think it is wrong code.

File: lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/Timers.java (hunk 3 above)
Comment (4e6) [reply lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/Timers.java:94]: MDN specifies a minimum delay for browsers (4ms) https://developer.mozilla.org/en-US/docs/Web/API/setInterval#delay_restrictions

Node.js specifies both minimum and maximum delay values https://nodejs.org/api/timers.html#setintervalcallback-delay-args

File: lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/Timers.java (hunk 2 above)
Comment (4e6) [reply lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/Timers.java]: I had concurrent execution errors in the JS code before. Maybe with a single executor, it won't be the case anymore. I will test it.

Comment (JaroslavTulach) [reply lib/java/ydoc-server/src/main/java/org/enso/ydoc/polyfill/web/Timers.java]: When there were multiple threads, it was indeed essential to return back to the original thread before executing the JavaScript function. However, now, when there is just a _single thread scheduling executor_ everything can be much simpler.

**Pull request merged**